File: focused_web_search.js

console.log("Starting ...");
var search_term = process.argv[2];
console.log("Search term from command line: " + search_term);


var links   = new Array;
var http    = require("http");
var request = require("request");
var sleep   = require("sleep");
var fs      = require('fs');


function debugResponse(res, body) {
	console.log("    res = |" + console.dir(res) +"|");
	console.log("    res.request = |" + Object.keys(res.request) +"|");
	console.log("    res.statusCode = " + res.statusCode);
	console.log("    res.request.uri.href = |" + res.request.uri.href +"|");
	console.log("      body = " + body);
}


function handleFileLinks(links) {
	var re = new RegExp(search_term);
	// console.log("##### handleFileLinks - links = " + links);

	for(var i=0; i < links.length; i++) {
		console.log("  checking link: |" + links[i] + "|");
		sleep.sleep(1);

		request(links[i], function(error, res, body) {
			// debugResponse(res, body);
 			var reArray = re.exec(body);

			if(reArray) {
				console.log("  Found Match: " + res.request.uri.href);
			} else {
				// console.log("  Not Found");
			}
		}).on('error', function(e) {
			console.log("Error: " + e.message);
		});
	}
}


fs.readFile('links.txt', 'utf-8', function(err, data) {
	console.log("Parsing file ...");
	if(err) {
		return console.log(err);
	}

	// console.log("data = " + data);
	var lines = data.split("\n");
	// console.log("lines = " + lines);

	for(var i=0; i < lines.length; i++) {
		var line = lines[i];
		// console.log("  Current Line: " + line);
		if(line.trim().length > 0) {
			// found content
			links[links.length] = line.trim();
		}
	}

	console.log("Done.");

        console.log("Evaluating links ...");
	handleFileLinks(links);
});

File: focused_web_search.exs


defmodule FocusedWebSearch do

  def search_links(web_links, args) do
    IO.puts "Searching urls ..."
    [search_term | _rest] = args
    :inets.start
    # :httpc.request(:get, {'http://www.erlang.org', []}, [], [])
    Enum.map(web_links, fn link -> 
      IO.puts "Calling link: #{link}"
      {:ok,{_status,_headers,content}} = :httpc.request('#{link}')
      # IO.puts "Content: #{content}   search_term = #{search_term}"
      if Regex.match?(~r/#{search_term}/, "#{content}") do
        IO.puts "!!!!! FOUND !!!!!"
      end
    end)
  end


  def process_file_contents(contents) do
    # IO.puts("Processing file contents #{contents}\n\n")
    lines = String.split(contents, "\n")
    # Enum.map(lines, fn l -> 
    #                      IO.puts "Line: #{l}"
    #                 end
    # )

    # ignore blank lines
    non_empty_lines = lines |> Stream.filter(fn l -> String.length(String.strip(l)) > 0 end) |> Enum.to_list

    search_links(non_empty_lines, System.argv())

    # IO.puts "Non-empty lines: #{List.to_string non_empty_lines}"
    # IO.puts "Non-empty lines:  " <> List.to_string(non_empty_lines)
    # Enum.map(non_empty_lines, fn l2 -> IO.puts "Non-empty line: #{l2}" end)
  end 
end


IO.puts "Starting ..."

filename = "links.txt"
IO.puts "Opening file: #{filename}"
case File.read(filename) do
  {:ok, body} -> FocusedWebSearch.process_file_contents(body)
  {:error, reason} -> IO.puts "Error: #{reason}"
end

# IO.puts "Command line arguments: " <> args

IO.puts "Done."



